biobalm/_pint_reachability.py

```diff
@@ -0,0 +1,157 @@
+        return False  # Cannot reach a stat in an empty set.
+
+    # Build a Pint model through an automata network and copy
+    # over the initial condition.
+    pint_model = InMemoryModel(_petri_net_as_automata_network(petri_net))
+    for var, level in initial_state.items():
+        pint_model.initial_state[var] = level
+
+    goal = _pint_build_symbolic_goal(target_states)
+
+    def failed(_model, _goal):
+        raise RuntimeError("Cannot verify.")
+
+    return pint_model.reachability(goal=goal, fallback=failed)  # type: ignore
+
+
+def _pint_build_symbolic_goal(states: Bdd) -> Goal:
+    """
+    A helper method which (very explicitly) converts a set of states
+    represented through a BDD into a Pint `Goal`.
+
+    Note that if `GOAL_SIZE_LIMIT` is exceeded, a partial goal is returned
+    that may not cover all the states in the argument `Bdd`. This avoids
+    exceeding the argument list size limit, but introduces additional source
+    of incompleteness into the reachability process.
+    """
+    assert not states.is_false()
+
+    goals: list[Goal] = []
+    limit = GOAL_SIZE_LIMIT
+    for clause in _optimized_recursive_dnf_generator(states):
+        named_clause = {
+            states.__ctx__().get_variable_name(var): int(value)
+            for var, value in clause.items()
+        }
+
+        limit -= len(named_clause)
+        if limit < 0:
+            # If the goal is too large to be passed as a command line argument,
+            # break here and don't continue. This is not ideal but I'm not sure
+            # how to fix this other than modifying `pint` itself.
+            if biobalm.succession_diagram.DEBUG:
+                print(
+                    "WARNING: `pint` goal size limit exceeded. A partial goal is used."
+                )
+            break
+
+        goal_atoms = [f"{var}={level}" for var, level in named_clause.items()]
+        goals.append(Goal(",".join(goal_atoms)))
+
+    return reduce(lambda a, b: a | b, goals)
+
+
+def _petri_net_as_automata_network(petri_net: DiGraph) -> str:
+    """
+    Takes a Petri net which was created by implicant encoding from a Boolean network,
+    and builds an automata network file (`.an`) compatible with the Pint tool.
+    """
+    automata_network = ""
+
+    # Go through all PN places and save them as model variables.
+    variable_set: set[str] = set()
+    for place, kind in petri_net.nodes(data="kind"):  # type: ignore
+        if kind != "place":
+            continue
+        variable_set.add(place_to_variable(place)[0])  # type: ignore[reportUnknownArgumentType] # noqa
+    variables = sorted(variable_set)
+
+    # Declare all variables with 0/1 domains.
+    for var in variables:
+        automata_network += f'"{var}" [0, 1]\n'
+
+    for transition, kind in petri_net.nodes(data="kind"):  # type: ignore
+        if kind != "transition":
+            continue
+
+        predecessors = set(petri_net.predecessors(transition))  # type: ignore
```

Comment:
These are sets of strings, right? I think if we just cast here we can remove the type ignores in this function. Maybe we can at least annotate as `predecessors: set[str] = ...` if we don't want to bother with casting.

Reply:
Makes sense. Will update.